internal/ota-api/dto.go

```diff
@@ -155,14 +158,58 @@ func (r OtaStatusDetail) String() string {
 		t = table.New()
 		t.SetHeader("Time", "Status", "Detail")
 		for _, s := range r.Details {
-			t.AddRow(formatHumanReadableTs(s.Timestamp), upperCaseFirst(s.State), s.StateData)
+			stateData := formatStateData(s.State, s.StateData, r.FirmwareSize, hasReachedFlashState(r.Details))
+			t.AddRow(formatHumanReadableTs(s.Timestamp), upperCaseFirst(s.State), stateData)
 		}
 		output += "\nDetails:\n" + t.Render()
 	}
 
 	return output
 }
 
+func hasReachedFlashState(states []State) bool {
+	for _, s := range states {
+		if s.State == "flash" {
+			return true
+		}
+	}
+	return false
+}
+
+func formatStateData(state, data string, firmware_size *int64, hasReceivedFlashState bool) string {
+	if data == "" {
+		return ""
+	}
+	if state == "fetch" {
+		// This is the state 'fetch' of OTA progress. This contains a number that represents the number of bytes fetched
+		if hasReceivedFlashState {
+			return buildSimpleProgressBar(float64(100))
+		}
+		if strings.ToLower(data) == "unknown" {
+			return ""
+		}
+		actualDownloadedData, err := strconv.Atoi(data)
+		if err != nil || firmware_size == nil || actualDownloadedData <= 0 || *firmware_size <= 0 { // Sanitize and avoid division by zero
```

Comment:
IMHO it will simple to read and reason about if you remove the pointer and use the zero value instead.
```suggestion
		if err != nil || actualDownloadedData <= 0 || firmware_size <= 0 { // Sanitize and avoid division by zero
```

Reply:
see changes